File: archive/python-legacy/polymarket/backtesting/market_simulator.py
"""
Market simulator for backtesting
Generates realistic simulated Polymarket weather markets from actual weather data
"""

import logging
import numpy as np
import pandas as pd
from typing import Dict, List, Tuple

logger = logging.getLogger(__name__)


def fahrenheit_to_celsius(f: float) -> float:
    return (f - 32) * 5 / 9


class MarketSimulator:
    """
    Generates simulated Polymarket weather prediction markets
    based on actual historical weather data for backtesting.
    """

    # Temperature thresholds in Fahrenheit (as Polymarket uses F)
    DEFAULT_TEMP_THRESHOLDS_F = [32, 50, 60, 70, 80, 90]

    def __init__(
        self,
        temp_thresholds_f: List[int] = None,
        market_noise_std: float = 0.10,
        seed: int = 42,
    ):
        """
        Args:
            temp_thresholds_f: Temperature thresholds in Fahrenheit for market generation
            market_noise_std: Std dev of noise added to climatological probability
                              to simulate market inefficiency
            seed: Random seed for reproducibility
        """
        self.temp_thresholds_f = temp_thresholds_f or self.DEFAULT_TEMP_THRESHOLDS_F
        self.market_noise_std = market_noise_std
        self.rng = np.random.RandomState(seed)

    def generate_markets(
        self,
        weather_data: pd.DataFrame,
        city: str,
        climatology: pd.DataFrame = None,
    ) -> pd.DataFrame:
        """
        Generate simulated prediction markets from actual weather data.

        For each day, creates markets for:
        - "Will max temperature exceed X°F?" for each threshold
        - "Will it rain today?"

        Args:
            weather_data: DataFrame with columns: date, temperature_max,
                         temperature_min, temperature_mean, precipitation_total
            city: City name (for market naming)
            climatology: Optional climatological averages for computing market prices.
                        If None, uses a rolling 30-day window from the data itself.

        Returns:
            DataFrame with columns: date, market_id, market_title, market_type,
                                   threshold, market_price, actual_outcome
        """
        if weather_data.empty:
            return pd.DataFrame()

        weather_data = weather_data.copy()
        weather_data["date"] = pd.to_datetime(weather_data["date"])
        weather_data = weather_data.sort_values("date").reset_index(drop=True)

        markets = []

        # Compute rolling climatological probabilities for market price generation
        clim_probs = self._compute_climatological_probs(weather_data)

        for idx, row in weather_data.iterrows():
            date = row["date"]
            temp_max_c = row.get("temperature_max")
            temp_mean_c = row.get("temperature_mean")
            precip = row.get("precipitation_total", 0)

            if pd.isna(temp_max_c):
                continue

            # Convert to Fahrenheit for market thresholds
            temp_max_f = temp_max_c * 9 / 5 + 32

            # Temperature threshold markets
            for threshold_f in self.temp_thresholds_f:
                actual_outcome = 1.0 if temp_max_f >= threshold_f else 0.0

                # Market price = climatological probability + noise
                clim_key = f"temp_above_{threshold_f}f"
                clim_prob = clim_probs.get(clim_key, {}).get(idx, 0.5)
                market_price = self._generate_market_price(clim_prob)

                markets.append({
                    "date": date,
                    "market_id": f"{city}_temp_{threshold_f}f_{date.strftime('%Y%m%d')}",
                    "market_title": f"Will {city} max temperature exceed {threshold_f}°F?",
                    "market_type": "temperature",
                    "threshold": threshold_f,
                    "market_price": market_price,
                    "actual_outcome": actual_outcome,
                    "city": city,
                })

            # Precipitation market
            if not pd.isna(precip):
                actual_rain = 1.0 if precip > 0.1 else 0.0  # >0.1mm = rain
                clim_prob = clim_probs.get("precipitation", {}).get(idx, 0.3)
                market_price = self._generate_market_price(clim_prob)

                markets.append({
                    "date": date,
                    "market_id": f"{city}_rain_{date.strftime('%Y%m%d')}",
                    "market_title": f"Will it rain in {city} today?",
                    "market_type": "precipitation",
                    "threshold": 0.1,
                    "market_price": market_price,
                    "actual_outcome": actual_rain,
                    "city": city,
                })

        df = pd.DataFrame(markets)
        logger.info(
            f"Generated {len(df)} simulated markets for {city} "
            f"over {weather_data['date'].nunique()} days"
        )
        return df

    def _compute_climatological_probs(
        self, weather_data: pd.DataFrame, window: int = 30
    ) -> Dict[str, Dict[int, float]]:
        """
        Compute rolling climatological probabilities for each market type.
        Uses a trailing window to avoid look-ahead bias.
        """
        probs = {}

        # Temperature thresholds
        if "temperature_max" in weather_data.columns:
            temp_max_f = weather_data["temperature_max"] * 9 / 5 + 32
            for threshold_f in self.temp_thresholds_f:
                key = f"temp_above_{threshold_f}f"
                exceeded = (temp_max_f >= threshold_f).astype(float)
                rolling_prob = exceeded.rolling(window=window, min_periods=7).mean()
                # Fill NaN at start with overall mean
                rolling_prob = rolling_prob.fillna(exceeded.mean())
                probs[key] = rolling_prob.to_dict()

        # Precipitation
        if "precipitation_total" in weather_data.columns:
            rained = (weather_data["precipitation_total"] > 0.1).astype(float)
            rolling_prob = rained.rolling(window=window, min_periods=7).mean()
            rolling_prob = rolling_prob.fillna(rained.mean())
            probs["precipitation"] = rolling_prob.to_dict()

        return probs

    def _generate_market_price(self, climatological_prob: float) -> float:
        """
        Generate a realistic market price from climatological probability.
        Adds noise to simulate market inefficiency and disagreement.
        """
        noise = self.rng.normal(0, self.market_noise_std)
        price = climatological_prob + noise
        # Clamp to valid probability range with some spread from 0/1
        return float(np.clip(price, 0.02, 0.98))

File: src/backtesting/market_simulator.rs
use std::collections::HashMap;

use rand::{Rng, SeedableRng};

use crate::types::{SimulatedMarket, WeatherRecord};
use crate::utils::clip;

pub fn fahrenheit_to_celsius(f: f64) -> f64 {
    (f - 32.0) * 5.0 / 9.0
}

#[derive(Debug, Clone)]
pub struct MarketSimulator {
    pub temp_thresholds_f: Vec<i32>,
    pub market_noise_std: f64,
    rng: rand::rngs::StdRng,
}

impl MarketSimulator {
    pub fn new(temp_thresholds_f: Option<Vec<i32>>, market_noise_std: f64, seed: u64) -> Self {
        Self {
            temp_thresholds_f: temp_thresholds_f.unwrap_or_else(|| vec![32, 50, 60, 70, 80, 90]),
            market_noise_std,
            rng: rand::rngs::StdRng::seed_from_u64(seed),
        }
    }

    pub fn generate_markets(&mut self, weather_data: &[WeatherRecord], city: &str) -> Vec<SimulatedMarket> {
        if weather_data.is_empty() {
            return Vec::new();
        }

        let mut sorted = weather_data.to_vec();
        sorted.sort_by_key(|r| r.date);
        let clim_probs = self.compute_climatological_probs(&sorted, 30);

        let mut markets = Vec::new();

        for (idx, row) in sorted.iter().enumerate() {
            let Some(temp_max_c) = row.temperature_max else {
                continue;
            };
            let temp_max_f = temp_max_c * 9.0 / 5.0 + 32.0;

            let thresholds = self.temp_thresholds_f.clone();
            for threshold in thresholds {
                let actual_outcome = if temp_max_f >= threshold as f64 { 1.0 } else { 0.0 };
                let key = format!("temp_above_{threshold}f");
                let clim_prob = clim_probs
                    .get(&key)
                    .and_then(|m| m.get(&idx))
                    .copied()
                    .unwrap_or(0.5);

                markets.push(SimulatedMarket {
                    date: row.date,
                    market_id: format!("{city}_temp_{threshold}f_{}", row.date.format("%Y%m%d")),
                    market_title: format!("Will {city} max temperature exceed {threshold}°F?"),
                    market_type: "temperature".to_string(),
                    threshold: threshold as f64,
                    market_price: self.generate_market_price(clim_prob),
                    actual_outcome,
                    city: city.to_string(),
                });
            }

            if let Some(precip) = row.precipitation_total {
                let actual_rain = if precip > 0.1 { 1.0 } else { 0.0 };
                let clim_prob = clim_probs
                    .get("precipitation")
                    .and_then(|m| m.get(&idx))
                    .copied()
                    .unwrap_or(0.3);

                markets.push(SimulatedMarket {
                    date: row.date,
                    market_id: format!("{city}_rain_{}", row.date.format("%Y%m%d")),
                    market_title: format!("Will it rain in {city} today?"),
                    market_type: "precipitation".to_string(),
                    threshold: 0.1,
                    market_price: self.generate_market_price(clim_prob),
                    actual_outcome: actual_rain,
                    city: city.to_string(),
                });
            }
        }

        markets
    }

    fn compute_climatological_probs(
        &self,
        weather_data: &[WeatherRecord],
        window: usize,
    ) -> HashMap<String, HashMap<usize, f64>> {
        let mut probs = HashMap::new();

        for threshold in &self.temp_thresholds_f {
            let key = format!("temp_above_{threshold}f");
            let exceeded: Vec<f64> = weather_data
                .iter()
                .map(|w| {
                    if w.temperature_max.map(|t| t * 9.0 / 5.0 + 32.0 >= *threshold as f64).unwrap_or(false)
                    {
                        1.0
                    } else {
                        0.0
                    }
                })
                .collect();
            probs.insert(key, rolling_probabilities(&exceeded, window, 7));
        }

        let rained: Vec<f64> = weather_data
            .iter()
            .map(|w| if w.precipitation_total.unwrap_or(0.0) > 0.1 { 1.0 } else { 0.0 })
            .collect();
        probs.insert(
            "precipitation".to_string(),
            rolling_probabilities(&rained, window, 7),
        );

        probs
    }

    fn generate_market_price(&mut self, climatological_prob: f64) -> f64 {
        let noise = rand_distr::Normal::new(0.0, self.market_noise_std)
            .ok()
            .map(|dist| self.rng.sample(dist))
            .unwrap_or(0.0);
        clip(climatological_prob + noise, 0.02, 0.98)
    }
}

fn rolling_probabilities(values: &[f64], window: usize, min_periods: usize) -> HashMap<usize, f64> {
    let overall = if values.is_empty() {
        0.5
    } else {
        values.iter().sum::<f64>() / values.len() as f64
    };

    let mut out = HashMap::new();
    for i in 0..values.len() {
        let start = i.saturating_sub(window.saturating_sub(1));
        let slice = &values[start..=i];
        let val = if slice.len() >= min_periods {
            slice.iter().sum::<f64>() / slice.len() as f64
        } else {
            overall
        };
        out.insert(i, val);
    }
    out
}
